src/Microsoft.Diagnostics.Monitoring.WebApi/CollectionRulePipelineState.cs

```diff
@@ -0,0 +1,209 @@
+            }
+        }
+
+        private void ActionExecutionFailed()
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState == CollectionRuleState.ActionExecuting);
+
+                CurrentState = CollectionRuleState.Running;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Running;
+            }
+        }
+
+        private void BeginThrottled()
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState != CollectionRuleState.Finished);
+
+                CurrentState = CollectionRuleState.Throttled;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Throttled;
+            }
+        }
+
+        private void EndThrottled()
+        {
+            lock (_lock)
+            {
+                if (CurrentState == CollectionRuleState.Throttled)
+                {
+                    CurrentState = CollectionRuleState.Running;
+                    CurrentStateReason = Strings.Message_CollectionRuleStateReason_Running;
+                }
+            }
+        }
+
+        public void StartupTriggerCompleted()
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState != CollectionRuleState.Finished);
+
+                CurrentState = CollectionRuleState.Finished;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Finished_Startup;
+            }
+        }
+
+        public void RuleDurationReached()
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState != CollectionRuleState.Finished);
+
+                CurrentState = CollectionRuleState.Finished;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Finished_RuleDuration;
+            }
+        }
+
+        public void ActionCountReached()
+        {
+            lock (_lock)
+            {
+                CurrentState = CollectionRuleState.Finished;
+                CurrentStateReason = Strings.Message_CollectionRuleStateReason_Finished_ActionCount;
+            }
+        }
+
+        public void RuleFailure(string errorMessage)
+        {
+            lock (_lock)
+            {
+                Debug.Assert(CurrentState != CollectionRuleState.Finished);
+
+                CurrentState = CollectionRuleState.Finished;
+                CurrentStateReason = string.Format(CultureInfo.InvariantCulture, Strings.Message_CollectionRuleStateReason_Finished_Failure, errorMessage);
+            }
+        }
+
+        public bool CanExecuteActions(DateTime currentTime)
+        {
+            lock (ExecutionTimestamps)
+            {
+                DequeueOldTimestamps(ExecutionTimestamps, ActionCountSlidingWindowDuration, currentTime);
+            }
+            bool canExecuteActions = !CheckForThrottling(ActionCountLimit, ActionCountSlidingWindowDuration, ExecutionTimestamps.Count);
```

Comment:
Should CheckForThrottling also be done under a lock?

Reply:
Discussed

Reply:
While the cloned version of the state can do this without locking, your _stateHolder object calls this method as well. The calls to ExecutionTimestamps.Count and the other properties should be under a lock.

Reply:
Maybe I misunderstood - when we talked about this on Friday, I thought we said we wouldn't need to lock around this since we're just getting the count for `ExecutionTimestamps` (and `ActionCountLimit` and `ActionCountSlidingWindowDuration` shouldn't change). Regardless, I can add that into the `lock`.